sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -89,6 +89,12 @@ object ConstantFolding extends Rule[LogicalPlan] {
           e
       }
 
+    // Don't replace ScalarSubquery if its plan is an aggregate that may suffer from a COUNT bug.
```

Comment:
I'm not a big fan of these band-aid fixes but I don't have a better idea. We should think about how to handle the count bug with a more general approach.

Reply:
I think the count bug is just a symptom, the main problem is that we handle decorrelation in 2 different places, and have the OptimizeSubqueries between them.